### ISSUE TITLE
[Epic] persist: Compaction

### ISSUE THREAD

**Issue by nmeagan11** [MaterializeInc/materialize#11491]
### Initiative and Theme

Materialize is Scalable; Materialize can maintain performance when data volumes increase

### Problem

Below are the use cases that compaction addresses:

1. **Cluster Health and Performance:** We want the amount of durable storage used and compute restart times to be proportional to the amount of unretracted data in a shard. At present, newly created consumers or readers of `persist` are required to hydrate state in time proportional to the full historical dataset, which will get slower over time without compaction. Specific examples of actions that will get slower over time include newly restarted `storaged` pods, new queries catching up, and running queries that don't maintain materialized state. We've even run into an example in practice where a basic operations become slower over time (see https://github.com/MaterializeInc/materialize/issues/13335).

2. **GDPR:** Being GDPR compliant requires us to delete data when requested to by a user. Fundamental to the design of STORAGE, sources are definite within a time interval `[since, upper)`. With compaction, for data at `time < since`, we will no longer retain data that has been removed. For example, if I insert `(1, data1)` at time `t1` and update that to `(1, data2)` at time `t2`, we will compact the stored data once `since > t2`, which removes all references to `data1`. Later on, we may have some additional work around providing a way for users to view the compaction boundary so they can confirm it's been reflected, and/or have a mechanism for the user to say they would like a compaction to occur at a specific time such that their delete is reflected. However, that is outside the scope of this epic.

3. **Decreased Spend in Money:** By compacting, we will have a smaller amount of data in S3, which will decrease storage costs (nice-to-have).

### Success Criteria

- As stated in https://github.com/MaterializeInc/materialize/issues/12847 and above, we aim to make the amount of durable storage used proportional to the amount of unretracted data in a shard, and make compute restart times proportional to the amount of unretracted data in a shard.
- A user should have acceptable cluster performance.

### Time Horizon

Large

#### Bugs:

 - [ ] https://github.com/MaterializeInc/materialize/issues/15093
   - I think the other issues are caused by this one :point_up: 
 - [ ] https://github.com/MaterializeInc/materialize/issues/15098
 - [ ] https://github.com/MaterializeInc/materialize/issues/15094
 - [ ] https://github.com/MaterializeInc/materialize/issues/15066
 - [ ] https://github.com/MaterializeInc/materialize/issues/14798 (maybe)

### Testing Tasks

- [ ] confirm the operation of compaction in practice with a dedicated test
- [ ] confirm that the desired restart time improvements have been achieved
- [x] verify that #13335 is fixed

**Comment by uce** [comment]
We can probably pick up the remaining testing tasks after the upsert source work has been implemented.

**Comment by philip-stoev** [comment]
QA sign-off currently blocked by various issues that show OOMs or compaction not kicking in under various scenarios.

**Comment by uce** [comment]
Closing this as the remaining work is tracked in #16607.
